### PR TITLE
buffer::raw: Conditional CRC cache

### PR DESCRIPTION
**Pull request by efirs**
Initialize and use CRC cache in buffer::raw only when the size of the buffer goes beyond threshold.
This shows 7% performance improvement in CPU bound 4K workload.

Signed-off-by: Evgeniy Firsov <evgeniy.firsov@sandisk.com>

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -135,24 +135,33 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
...
+			crc_lock = new RWLock("buffer::raw::crc_lock", false);
+			crc_map = new map<pair<size_t, size_t>, pair<uint32_t, uint32_t>>;
+		}
+	}
+    virtual ~raw() {
+		if(crc_lock)
+			delete crc_lock;
+		if(crc_map)
+			delete crc_map;
+	}
```
Comment (liewegas) [inline src/common/buffer.cc]: I would expect the allocation/free of the the lock and map to be pretty expensive... why not leave it in the struct?  The constructor for these is going to be basically free (zeroing a few fields), right?

Comment (liewegas) [inline src/common/buffer.cc]: Lots of style and whitespace issues.  Please use an editor that respects the modeline at the top of the file (it works with vim and emacs) and see the CodingStyle doc

**Comment by efirs** [comment]
@liewegas for the map it might be cheap, right. But mutex has a string member in it, which requires malloc and copy. So I tried to remove everything from the fast path.

**Comment by liewegas** [comment]
if we change it to a fixed-length array (that we do a search over, instead 
of a map) we can switch to a spinlock.

**Comment by liewegas** [comment]
@efirs Do you want to try updating this with a fixed-length array?  I think it'll show an even better performance improvement, and avoid the mutex allocation by letting us use a spinlock (with constant allocation/copy time).

**Comment by efirs** [comment]
@liewegas I'll do some more testing and see if I can remove mallocs and use static array.
But, as per my testing, jemalloc does a great job and mallocs are very inexpensive and shouldn't be of a big concern here.

**Pull request closed without merging**